dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/compute/functions/ServerWithNatRuleToNodeMetadata.java

```diff
@@ -0,0 +1,123 @@
+   private final BaseImageToHardware baseImageToHardware;
+   private final Map<String, Credentials> credentialStore;
+   private final DimensionDataCloudControlApi api;
+
+   @Inject
+   ServerWithNatRuleToNodeMetadata(@Memoized final Supplier<Set<? extends Location>> locations,
+         final GroupNamingConvention.Factory namingConvention, final BaseImageToImage baseImageToImage,
+         final BaseImageToHardware baseImageToHardware, final Map<String, Credentials> credentialStore,
+         final DimensionDataCloudControlApi api) {
+      this.nodeNamingConvention = checkNotNull(namingConvention, "namingConvention").createWithoutPrefix();
+      this.locations = checkNotNull(locations, "locations");
+      this.baseImageToImage = checkNotNull(baseImageToImage, "baseImageToImage");
+      this.baseImageToHardware = checkNotNull(baseImageToHardware, "osImageToHardware");
+      this.credentialStore = checkNotNull(credentialStore, "credentialStore cannot be null");
+      this.api = checkNotNull(api, "api cannot be null");
+   }
+
+   @Override
+   public NodeMetadata apply(final ServerWithExternalIp serverWithExternalIp) {
+      NodeMetadataBuilder builder = new NodeMetadataBuilder();
+      Server server = serverWithExternalIp.server();
+      builder.ids(server.id());
+      builder.name(server.name());
+      builder.hostname(serverWithExternalIp.server().description());
+      if (server.datacenterId() != null) {
+         builder.location(find(locations.get(), idEquals(nullToEmpty(server.datacenterId()))));
+      }
+      builder.group(nodeNamingConvention.groupInUniqueNameOrNull(server.name()));
+      BaseImage baseImage = getBaseImage(server.sourceImageId());
+      builder.hardware(baseImageToHardware.apply(baseImage));
+      Image image = baseImageToImage.apply(baseImage);
+      if (image != null) {
+         builder.imageId(image.getId());
+         builder.operatingSystem(image.getOperatingSystem());
+      }
+      if (server.state() != null) {
+         builder.status(serverStateToNodeStatus.get(server.state()));
+      }
+
+      String privateAddress = null;
+      if (server.networkInfo() != null && server.networkInfo().primaryNic() != null
+            && server.networkInfo().primaryNic().privateIpv4() != null) {
+         privateAddress = server.networkInfo().primaryNic().privateIpv4();
+         builder.privateAddresses(ImmutableSet.of(privateAddress));
+      }
+      if (privateAddress != null && serverWithExternalIp.externalIp() != null) {
+         builder.publicAddresses(ImmutableSet.of(serverWithExternalIp.externalIp()));
+      }
+
+      // DimensionData does not provide a way to get the credentials.
+      // Try to return them from the credential store
+      Credentials credentials = credentialStore.get("node#" + server.id());
+      if (credentials instanceof LoginCredentials) {
+         builder.credentials(LoginCredentials.class.cast(credentials));
+      }
```

Comment:
Remove this. The credentials are [already set by jclouds](https://github.com/jclouds/jclouds/blob/master/compute/src/main/java/org/jclouds/compute/strategy/impl/AdaptingComputeServiceStrategies.java#L89-L90).